### PR TITLE
Free session_ID, dhe_context and keys for: key exchange and finish

### PR DESCRIPTION
**Pull request by Wenxing-hou**
Fix: #476

 1) free session_id
 2) free dhe_context if key is not established
 3) zero negotiated keys

Signed-off-by: Wenxing Hou <wenxing.hou@intel.com>

### CODE REVIEW

File: library/spdm_requester_lib/libspdm_req_finish.c [hunk 1]
```diff
@@ -250,6 +266,12 @@ return_status try_spdm_send_receive_finish(IN spdm_context_t *spdm_context,
     spdm_context->error_state = LIBSPDM_STATUS_SUCCESS;
 
     return RETURN_SUCCESS;
+error:
```
Comment (jyao1) [inline library/spdm_requester_lib/libspdm_req_finish.c:269]: Not all error goes to this path. For example:
```
    status = spdm_send_spdm_request(spdm_context, &session_id,
                    spdm_request_size, &spdm_request);
    if (RETURN_ERROR(status)) {
        return RETURN_DEVICE_ERROR;
    }
```

File: library/spdm_requester_lib/libspdm_req_psk_finish.c [hunk 2]
```diff
@@ -163,6 +179,10 @@ return_status try_spdm_send_receive_psk_finish(IN spdm_context_t *spdm_context,
     spdm_context->error_state = LIBSPDM_STATUS_SUCCESS;
 
     return RETURN_SUCCESS;
+
+error:
```
Comment (jyao1) [inline library/spdm_requester_lib/libspdm_req_psk_finish.c:183]: No all error goes to this path. For example:
```
        status = spdm_handle_error_response_main(
            spdm_context, &session_id,
            &spdm_response_size, &spdm_response,
            SPDM_PSK_FINISH, SPDM_PSK_FINISH_RSP,
            sizeof(spdm_psk_finish_response_max_t));
        if (RETURN_ERROR(status)) {
            return status;
        }
```

File: library/spdm_requester_lib/libspdm_req_psk_exchange.c [hunk 3]
```diff
@@ -285,6 +285,10 @@ return_status try_spdm_send_receive_psk_exchange(
 
     ptr += measurement_summary_hash_size;
 
+    if ( spdm_response.opaque_length > SPDM_MAX_OPAQUE_DATA_SIZE) {
+        libspdm_free_session_id(spdm_context, *session_id);
+        return RETURN_SECURITY_VIOLATION;
+    }
```
Comment (jyao1) [inline library/spdm_requester_lib/libspdm_req_psk_exchange.c:291]: Some other places need fix:
```
    status = libspdm_append_message_k(spdm_context, session_info, TRUE, &spdm_request,
                       spdm_request_size);
    if (RETURN_ERROR(status)) {
        return RETURN_SECURITY_VIOLATION;
    }
```

**Comment by Wenxing-hou** [comment]
I have checked the code carefully and updated the code.  Thanks.

**Comment by jyao1** [comment]
Still have issue:

library/spdm_requester_lib/libspdm_req_finish.c
```
    if (spdm_response.header.request_response_code == SPDM_ERROR) {
        if (spdm_response.header.param1 != SPDM_ERROR_CODE_RESPONSE_NOT_READY) {
            libspdm_reset_message_f (spdm_context, session_info);
        }
        status = spdm_handle_error_response_main(
            spdm_context, &session_id,
            &spdm_response_size, &spdm_response,
            SPDM_FINISH, SPDM_FINISH_RSP,
            sizeof(spdm_finish_response_mine_t));
        if (RETURN_ERROR(status)) {
            return status;
        }
    } else if (spdm_response.header.request_response_code !=
```

library/spdm_requester_lib/libspdm_req_psk_finish.c
```
    if (spdm_response.header.request_response_code == SPDM_ERROR) {
        status = spdm_handle_error_response_main(
            spdm_context, &session_id,
            &spdm_response_size, &spdm_response,
            SPDM_PSK_FINISH, SPDM_PSK_FINISH_RSP,
            sizeof(spdm_psk_finish_response_max_t));
        if (RETURN_ERROR(status)) {
            return status;
        }
    } else if (spdm_response.header.request_response_code !=
```

**Comment by Wenxing-hou** [comment]
I think the follow code should be stayed.
```
    if (spdm_response.header.request_response_code == SPDM_ERROR) {
        status = spdm_handle_error_response_main(
            spdm_context, &session_id,
            &spdm_response_size, &spdm_response,
            SPDM_PSK_FINISH, SPDM_PSK_FINISH_RSP,
            sizeof(spdm_psk_finish_response_max_t));
        if (RETURN_ERROR(status)) {
            return status;
```
Because the status may return: RETURN_NO_RESPONSE. And we shouldn't free the session ID. We will try it again.
```
return_status spdm_send_receive_psk_finish(IN spdm_context_t *spdm_context,
                       IN uint32_t session_id)
{
    uintn retry;
    return_status status;

    retry = spdm_context->retry_times;
    do {
        status = try_spdm_send_receive_psk_finish(spdm_context,
                              session_id);
        if (RETURN_NO_RESPONSE != status) {
            return status;
        }
    } while (retry-- != 0);

    return status;
}
```

**Comment by jyao1** [comment]
looks good now.

**Review by steven-bellock — APPROVED** [review]

**Pull request merged**